presto-spi/src/main/java/com/facebook/presto/spi/plan/SetOperationNode.java

```diff
@@ -0,0 +1,116 @@
+package com.facebook.presto.spi.plan;
+
+import com.facebook.presto.spi.relation.VariableReferenceExpression;
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+import javax.annotation.concurrent.Immutable;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+
+import static java.lang.String.format;
+import static java.util.Collections.unmodifiableList;
+import static java.util.Collections.unmodifiableMap;
+import static java.util.Objects.requireNonNull;
+import static java.util.stream.Collectors.toList;
+
+@Immutable
+public abstract class SetOperationNode
+        extends PlanNode
+{
+    private final List<PlanNode> sources;
+    private final LinkedHashMap<VariableReferenceExpression, List<VariableReferenceExpression>> outputToInputs;
+    private final List<VariableReferenceExpression> outputVariables;
+
+    @JsonCreator
+    protected SetOperationNode(
+            @JsonProperty("id") PlanNodeId id,
+            @JsonProperty("sources") List<PlanNode> sources,
+            @JsonProperty("outputToInputs") LinkedHashMap<VariableReferenceExpression, List<VariableReferenceExpression>> outputToInputs)
+    {
+        super(id);
+
+        requireNonNull(sources, "sources is null");
+        checkArgument(!sources.isEmpty(), "Must have at least one source");
+        requireNonNull(outputToInputs, "outputToInputs is null");
+
+        this.sources = unmodifiableList(new ArrayList<>(sources));
+        this.outputToInputs = new LinkedHashMap<>(outputToInputs);
+        this.outputVariables = unmodifiableList(new ArrayList<>(outputToInputs.keySet()));
+
+        for (Collection<VariableReferenceExpression> inputs : this.outputToInputs.values()) {
+            checkArgument(
+                    inputs.size() == this.sources.size(),
+                    format("Every source needs to map its symbols to an output %s operation symbol", this.getClass().getSimpleName()));
+        }
+
+        // Make sure each source positionally corresponds to their Symbol values in the Multimap
+        for (int i = 0; i < sources.size(); i++) {
+            for (List<VariableReferenceExpression> expectedInputs : this.outputToInputs.values()) {
+                checkArgument(sources.get(i).getOutputVariables().contains(expectedInputs.get(i)), "Source does not provide required symbols");
+            }
+        }
+    }
+
+    @Override
+    @JsonProperty
+    public List<PlanNode> getSources()
+    {
+        return sources;
+    }
+
+    @JsonProperty
+    public LinkedHashMap<VariableReferenceExpression, List<VariableReferenceExpression>> getVariableMapping()
+    {
+        return outputToInputs;
+    }
+
+    @Override
+    public List<VariableReferenceExpression> getOutputVariables()
+    {
+        return outputVariables;
+    }
+
+    public List<VariableReferenceExpression> sourceOutputLayout(int sourceIndex)
```

Comment:
Since this is not utility function uses `getVariableMapping` and `getOutputVariable`, probably just move this method into `SetOperationNodeUtil`? same for `sourceVariableMap`?